### PR TITLE
[react-dom] Update types for latest state of Float

### PR DESCRIPTION
**Pull request by eps1lon**
1. https://github.com/facebook/react/pull/26237
2. https://github.com/facebook/react/pull/27220
3. https://github.com/facebook/react/pull/26880
4. https://github.com/facebook/react/pull/26939
5. https://github.com/facebook/react/pull/26940
6. https://github.com/facebook/react/pull/27096
7. Concrete string literal for `crossOrigin`

@gnoff Could you review this one?

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react-dom ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-dom@latest/))
was missing the following properties:
1. createRoot
1. hydrateRoot
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d38fd3d12f7410e01f1e0cf5f016d266162f0fe8
</p>

**Review by rbuckton — APPROVED** [review]

File: types/react-dom/canary.d.ts [hunk 1]
```diff
@@ -32,21 +32,64 @@ import ReactDOM = require(".");
...
+        // Don't create a helper type.
+        // It would have to be in module scope to be inlined in TS tooltips.
+        // But then it becomes part of the public API.
+        // TODO: Upstream to microsoft/TypeScript-DOM-lib-generator -> w3c/webref
+        // since the spec has a notion of a dedicated type: https://html.spec.whatwg.org/multipage/urls-and-fetching.html#cors-settings-attribute
+        crossOrigin?: "anonymous" | "use-credentials" | "" | undefined;
+    }
+    function preconnect(href: string, options?: PreconnectOptions): void;
+
+    type PreloadAs = "font" | "image" | "script" | "style";
```
Comment (gnoff) [inline types/react-dom/canary.d.ts:47]: Internally we just treat this as a string argument now however what probably make sense in TS is to enumerate all of the known valid as types for preloads

```
// from: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/preload#what_types_of_content_can_be_preloaded

audio
document
embed
fetch
object
script
style
track
worker
video
```

File: types/react-dom/canary.d.ts [hunk 2]
```diff
@@ -32,21 +32,64 @@ import ReactDOM = require(".");
...
 
+    // https://html.spec.whatwg.org/multipage/links.html#link-type-modulepreload
+    type PreloadModuleAs = RequestDestination;
+    interface PreloadModuleOptions {
+        /**
+         * @default "script"
+         */
+        as: PreloadModuleAs;
+        crossOrigin?: "anonymous" | "use-credentials" | "" | undefined;
+        integrity?: string | undefined;
```
Comment (gnoff) [inline types/react-dom/canary.d.ts:69]:
```suggestion
        integrity?: string | undefined;
        nonce?: string | undefined;
```

File: types/react-dom/canary.d.ts [hunk 3]
```diff
@@ -32,21 +32,64 @@ import ReactDOM = require(".");
...
+
+    // Will be expanded to include all of https://github.com/tc39/proposal-import-attributes
+    type PreinitModuleAs = "script";
+    interface PreinitModuleOptions {
+        /**
+         * @default "script"
+         */
+        as?: PreinitModuleAs;
+        crossOrigin?: "anonymous" | "use-credentials" | "" | undefined;
+        integrity?: string | undefined;
```
Comment (gnoff) [inline types/react-dom/canary.d.ts:92]:
```suggestion
        integrity?: string | undefined;
        nonce?: string | undefined;
```

**Review by gnoff — CHANGES_REQUESTED** [review]

**Review by gnoff — APPROVED** [review]

**Pull request merged**